L1Trigger/Phase2L1GT/plugins/L1GTBoardWriter.cc

```diff
@@ -0,0 +1,104 @@
+
+private:
+  void analyze(const edm::Event&, const edm::EventSetup&) override;
+  void endJob() override;
+
+  const std::vector<unsigned int> channels_;
+  const std::vector<unsigned long long> algoBitMask_;
+  const edm::EDGetTokenT<P2GTAlgoBlockCollection> algoBlocksToken_;
+  l1t::demo::BoardDataWriter boardDataWriter_;
+};
+
+L1GTBoardWriter::L1GTBoardWriter(const edm::ParameterSet& config)
+    : channels_(config.getParameter<std::vector<unsigned int>>("channels")),
+      algoBitMask_(config.exists("algoBitMask") ? config.getParameter<std::vector<unsigned long long>>("algoBitMask")
+                                                : std::vector<unsigned long long>()),
+      algoBlocksToken_(consumes<P2GTAlgoBlockCollection>(config.getParameter<edm::InputTag>("algoBlocksTag"))),
+      boardDataWriter_(
+          config.exists("patternFormat") ? l1t::demo::parseFileFormat(config.getParameter<std::string>("patternFormat"))
+                                         : l1t::demo::FileFormat::EMPv2,
+          config.getParameter<std::string>("outputFilename"),
+          9,
+          1,
+          config.exists("maxLines") ? config.getParameter<unsigned int>("maxLines") : 1024,
+          [](const std::vector<unsigned int>& channels) {
+            l1t::demo::BoardDataWriter::ChannelMap_t channelMap;
+            for (unsigned int channel : channels) {
+              channelMap.insert({l1t::demo::LinkId{"Algos", channel}, {l1t::demo::ChannelSpec{1, 0, 0}, {channel}}});
+            }
+            return channelMap;
+          }(channels_)) {}
+
+void L1GTBoardWriter::analyze(const edm::Event& event, const edm::EventSetup& iSetup) {
+  l1t::demo::EventData eventData;
+  const P2GTAlgoBlockCollection& algoBlocks = event.get(algoBlocksToken_);
+
+  auto algoBlockIt = algoBlocks.begin();
+  auto algoMaskIt = algoBitMask_.begin();
+
+  for (unsigned int channel : channels_) {
+    std::vector<ap_uint<64>> bits(9, 0);
+    for (std::size_t word = 0; word < 9; word++) {
+      ap_uint<64> mask = algoMaskIt != algoBitMask_.end() ? *algoMaskIt++ : ~static_cast<unsigned long long>(0);
+
+      for (std::size_t idx = 0; idx < 64 && algoBlockIt != algoBlocks.end(); idx++) {
+        bits[word].set(idx, algoBlockIt->decisionBeforeBxMaskAndPrescale() && mask.bit(idx));
+        algoBlockIt++;
+      }
+    }
+
+    eventData.add({"Algos", channel}, bits);
+  }
+
+  boardDataWriter_.addEvent(eventData);
+}
+
+void L1GTBoardWriter::endJob() { boardDataWriter_.flush(); }
+
+void L1GTBoardWriter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<std::string>("outputFilename");
+  desc.add<edm::InputTag>("algoBlocksTag");
+  desc.add<std::vector<unsigned int>>("channels");
+  desc.addOptional<std::vector<unsigned long long>>("algoBitMask");
+  desc.addOptional<unsigned int>("maxLines", 1024);
+  desc.addOptional<std::string>("patternFormat");
```

Comment:
The constructor (initializer list) uses default values for all of these. Why making these as required parameters with the default values (that would avoid the `exists()` calls)?